icu4c/source/i18n/complexunitsconverter.cpp

```diff
@@ -0,0 +1,108 @@
+// © 2020 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+#include "unicode/utypes.h"
+
+#if !UCONFIG_NO_FORMATTING
+
+#include <math.h>
+#include <utility>
+
+#include "cmemory.h"
+#include "complexunitsconverter.h"
+#include "uassert.h"
+#include "unicode/fmtable.h"
+#include "unitconverter.h"
+
+U_NAMESPACE_BEGIN
+
+ComplexUnitsConverter::ComplexUnitsConverter(const MeasureUnit inputUnit, const MeasureUnit outputUnits,
+                                             const ConversionRates &ratesInfo, UErrorCode &status) {
+
+    if (outputUnits.getComplexity(status) != UMeasureUnitComplexity::UMEASURE_UNIT_MIXED) {
+        unitConverters_.emplaceBackAndConfirm(status, inputUnit, outputUnits, ratesInfo, status);
+        units_.emplaceBackAndConfirm(status, outputUnits);
+        return;
+    }
+
+    // In case the `outputUnits` are `UMEASURE_UNIT_MIXED` such as `foot+inch`. In this case we need more
+    // converters to convert from the `inputUnit` to the first unit in the `outputUnits`. Then, a
+    // converter from the first unit in the `outputUnits` to the second unit and so on.
+    //      For Example:
+    //          - inputUnit is `meter`
+    //          - outputUnits is `foot+inch`
+    //              - Therefore, we need to have two converters:
+    //                      1. a converter from `meter` to `foot`
+    //                      2. a converter from `foot` to `inch`
+    //          - Therefore, if the input is `2 meter`:
+    //              1. convert `meter` to `foot` --> 2 meter to 6.56168 feet
+    //              2. convert the residual of 6.56168 feet (0.56168) to inches, which will be (6.74016
+    //              inches)
+    //              3. then, the final result will be (6 feet and 6.74016 inches)
+    int32_t length;
+    auto singleUnits = outputUnits.splitToSingleUnits(length, status);
+    MaybeStackVector<MeasureUnit> singleUnitsInOrder;
+    for (int i = 0; i < length; ++i) {
+        // TODO(younies): ensure units being in order in phase 2. Now, the units in order by default.
+        singleUnitsInOrder.emplaceBackAndConfirm(status, singleUnits[i]);
+    }
```

Comment:
All of this can be simplified to (including sorting):

```suggestion
    MaybeStackVector<SingleUnitImpl> singleUnitsInOrder =
        MeasureUnitImpl::forMeasureUnitMaybeCopy(outputUnits, status).units;
    uprv_sortArray(
        singleUnitsInOrder.getAlias(),
        singleUnitsInOrder.length(),
        sizeof(singleUnitsInOrder[0]),
        compareSingleUnits,
        nullptr,
        false,
        &status);
```

You may need to add `compareSingleUnits` to `measunit_impl.h`.  It is currently implemented in `measunit_extra.cpp`.  Alternatively, I think this will sort the units in descending order, so maybe you want to write a custom sorting function instead (the reverse of what `compareSingleUnits` does).

Reply:
I have added a TODO right now, could I do this after submitting this PR ?

Reply:
I don't see your new TODO, only the old TODO that says "Phase 2".  I think we should fix this before Phase 2.

Reply:
I removed the stage2 part.

Reply:
with a hint